File: CSE2018/chap9/sources/stack.c
#include <stdio.h>
#include <stdlib.h>
#include "../headers/node.h"

/* Define 'stack' using typedef and struct.
	stack has top, size */
typedef struct {
	int size;
	struct node* top;
} stack;

/* Prototypes */
stack* init_stack();
void push(stack*, node*);
void pop(stack*);
int top(stack*);
void print_stack(stack*);

int main(int argc, char const *argv[])
{
	int i;
	stack* s = init_stack();
	push(s, newnode(1));
	printf("top : %d\n", top(s));
	pop(s);
	pop(s);
	for (i = 2; i < 6; i++) {
		push(s, newnode(i));
	}
	pop(s);
	print_stack(s);
	for (i = 0; i < 3; i++) pop(s);
	top(s);
	return 0;
}

stack* init_stack() {
	stack* temp = (stack*) malloc(sizeof(stack));

	temp->size = 0;
	temp->top = NULL;

	return temp;
}

void push(stack* stack, node* newnode) {
	stack->size++;
	if (stack->top == NULL) {
		stack->top = newnode;
		return;
	}
	newnode->next = stack->top;
	stack->top = newnode;
}

void pop(stack* stack) {
	if (stack->top == NULL) {
		printf("%s\n", "pop(): stack is empty.");
		return;
	}
	node* temp = stack->top;
	stack->top = stack->top->next;
	free(temp);
	stack->size--;
}

int top(stack* stack) {
	if (stack->top == NULL) {
		printf("%s\n", "top(): stack is empty.");
		return 0;
	}
	return stack->top->val;
}

void print_stack(stack* stack) {
	node* iter = stack->top;

	printf("%s\n", "TOP");
	for (int i = 1; i <= stack->size; i++, iter = iter->next) {
		printf("[%d]\n", iter->val);
	}
}

File: CSE2010/06/Stack.java
public class Stack {       
    private Node top; 
    private class Node {
        private String data;
        private Node next;
        Node(String data){
         this.data = data;
        }
    }
    
    // Constructor (생성자)
    public Stack() {
        top = null;
    }

    /*
     * Stack 안에 있는 elements 의 개수를 반환한다.
     */
    public int size(){
        int n = 0;
        Node temp = top;

        while (temp != null) {
            temp = temp.next;
            n++;
        }

        return n;
    }

    /*
     * Stack이 비어있는지를 반환한다.
     */
    public boolean isEmpty(){
        if (top == null) {
            return true;
        }
        else {
            return false;
        }
    }

    /*
     * Stack의 top 자리에 data를 가지는 Node를 삽입한다.
     */
    public void push(String data){
        Node temp = new Node(data);
        temp.next = top;
        top = temp;
    }
   
    /*
     * Stack의 top을 반환한다. (제거 x)
     */    
    public String top() {
        if (top == null) {
            System.out.println("There is no data.");

            return null;
        }
        System.out.println("Top data = " + top.data);

        return top.data;
    }


    /*
     * Stack의 top을 제거하며, 제거한 Node를 반환한다.
     */    
    public Node pop(){
        if (top == null) {
            System.out.println("There is no data.");

            return null;
        }
        Node temp = top;
        top = top.next;
        
        return temp;
    }
   

    /*
     * Stack 을 top부터 마지막 Node까지 출력한다.
     */    
    public void printStack(){
        Node temp = top;

        while (temp != null) {
            System.out.print(temp.data + " ");
            temp = temp.next;
        }
        System.out.println();
    }


}
